04 - Functional Programming/04 - Functional Programming.ipynb

```diff
@@ -0,0 +1,3551 @@
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "RH0Pc07HD4HQ",
+        "outputId": "b1bce3a4-23b4-4ed7-c321-644216299fd4"
+      },
+      "execution_count": 22,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "[('Иван', 22), ('Любо', 35), ('Мария', 35), ('Митко', 51), ('Георги', 71)]\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Lazy evaluation"
+      ],
+      "metadata": {
+        "id": "gkVl1LVhC3mE"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Съществуват различни стратегии за оценяване (или изчисляване) на изрази - най-позната е т.нар. \"нетърпеливо оценяване\" или \"оценяване на момента\" - целия израз се изчислява на момента.\n",
+        "\n",
+        "Нека е даден израза `a + b + c`. Първо, ще бъде оценена стойността на `a`, после на `b`. Следващата стъпка е да се извърши операцията събиране. Резултатът от първата операция ще бъде използван за следващата операция събиране, заедно с `c`. \n",
+        "\n",
+        "Пример за оценяване на момента може да бъде следната функция:"
+      ],
+      "metadata": {
+        "id": "BmDBNoJpnShQ"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "def f(x):\n",
+        "    print(f'x = {x}')\n",
+        "    return x\n",
+        "\n",
+        "print(f(1) + f(2))"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "mfuEbmgqomQQ",
+        "outputId": "22474686-0b7e-4a49-ae30-dfe0a7c124ae"
+      },
+      "execution_count": 23,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "x = 1\n",
+            "x = 2\n",
+            "3\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "При оценяването на момента (eager evaluation), всички необходими данни за извършване на изчислението (променливи, функции, др.) се зареждат наведнъж в паметта. Освен това, всички изчисления трябва да бъда извършени точно на момента."
+      ],
+      "metadata": {
+        "id": "BAo7aBEQowo3"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Тук се намесва т.нар. \"мързеливо\" оценяване - изчисленията се извършват само при необходимост. Това ни позволява да си спестим някои изчисления (както и да работим с т.нар. \"безкрайни\" колекции).\n",
+        "\n",
+        "Най-близкото до \"мързеливо\" оценяване, което познаваме е изчисляването на логическите изрази при `if` блокове и други булеви изрази - там, всеки елемент от израза се изчислява, при условие че предишния израз не е достатъчен за оценяване на целия израз (т.е. ако имаме един израз който е със стойност на `True`, тези след него не се оценяват). \n",
```

Comment:
This doesn't seem entirely correct. It's true only in the case of an OR expression. Examples:

```python
def true():
    print("I'm lazy")
    return True
True or true()
```
Here `foo()` is never called.

```python
def true():
    print("I'm lazy, but I'll run this time")
    return True
False or true()
```
Here `foo()` is called.

When we talk about `and`, however, the situations will be reversed.

```python
def true():
    print("I'm lazy, but I'll run this time")
    return True

# Short circuit
False and true()
# We have to evaluate everything to determine the result
True and true()
```

Reply:
Reworded a bit.
As a note: Short-circuit != lazy evaluation

Reply:
Yeah, I used the wrong word here